### PR TITLE
fix a sqlite bug

### PR DESCRIPTION
**Pull request by ringabout**
## Case One(The bug)
Before:
```nim
import std/[db_sqlite, os]


block:
  var db = db_sqlite.open("db.sqlite3", "", "", "")
  try:
    db.exec(sql("CREATE TABLE table1 (url TEXT, other_field INT);"))
    db.exec(sql("REPLACE INTO table (url, another_field) VALUES (?, '123');"))
  except DbError as e:
    doAssert e.msg == "The number of parameter doesn't match that of \"?\" "
  finally:
    db.close()
    removeFile("db.sqlite3")
```

raise IndexDefect

```
C:\Users\blue\Documents\GitHub\Nim\test6.nim(8) test6
C:\Users\blue\.choosenim\toolchains\nim-#devel\lib\impure\db_sqlite.nim(278) exec
C:\Users\blue\.choosenim\toolchains\nim-#devel\lib\impure\db_sqlite.nim(243) tryExec
C:\Users\blue\.choosenim\toolchains\nim-#devel\lib\impure\db_sqlite.nim(218) dbFormat
C:\Users\blue\.choosenim\toolchains\nim-#devel\lib\system\fatal.nim(53) sysFatal
Error: unhandled exception: index out of bounds, the container is empty [IndexDefect]
```

## Case Two(silent ignored bug)

When the number of parameter exceeds, still no raises like before.
```nim
block:
  var db = db_sqlite.open("db.sqlite3", "", "", "")
  try:
    db.exec(sql("CREATE TABLE table1 (url TEXT, other_field INT);"))
    db.exec(sql("INSERT INTO table1 (url, other_field) VALUES (?, ?);"), "http://domain.com/test?param=1", 123, 134)
  finally:
    db.close()
    removeFile("db.sqlite3")
```

But it can be fixed later(it is controversial to be fixed in this PR)

add a line before `res` variable
```nim
  if a != args.len:
    dbError("""The number of parameter doesn't match that of "?" """)
```

## future work
- [ ] https://github.com/nim-lang/Nim/pull/18669/files#r686051480

### CODE REVIEW

File: lib/std/private/dbutils.nim [hunk 1]
```diff
@@ -0,0 +1,15 @@
+import db_common
+
+
+template dbFormatImpl*(formatstr: SqlQuery, dbQuote: proc (s: string): string, args: varargs[string]): string =
```
Comment (ringabout) [inline lib/std/private/dbutils.nim:4]: Every `db_*` defines its own dbFormat. I will refactor others later.

For example

## db_odbc
```nim
proc dbFormat(formatstr: SqlQuery, args: varargs[string]): string {.
                  noSideEffect.} =
  ## Replace any `?` placeholders with `args`,
  ## and quotes the arguments
  result = ""
  var a = 0
  for c in items(string(formatstr)):
    if c == '?':
      add(result, dbQuote(args[a]))
      inc(a)
    else:
      add(result, c)
```

## db_postgres
IMO not correct, should be still indexDefect when the number doesn't match(I haven't checked it)
```nim
proc dbFormat(formatstr: SqlQuery, args: varargs[string]): string =
  result = ""
  var a = 0
  if args.len > 0 and not string(formatstr).contains("?"):
    dbError("""parameter substitution expects "?" """)
  if args.len == 0:
    return string(formatstr)
  else:
    for c in items(string(formatstr)):
      if c == '?':
        add(result, dbQuote(args[a]))
        inc(a)
      else:
        add(result, c)
```

## db_mysql

```nim
proc dbFormat(formatstr: SqlQuery, args: varargs[string]): string =
  result = ""
  var a = 0
  for c in items(string(formatstr)):
    if c == '?':
      add(result, dbQuote(args[a]))
      inc(a)
    else:
      add(result, c)
```

File: lib/impure/db_sqlite.nim [hunk 2]
```diff
@@ -642,7 +635,7 @@ proc getValue*(db: DbConn,  stmtName: SqlPrepared): string
 
 proc tryInsertID*(db: DbConn, query: SqlQuery,
                   args: varargs[string, `$`]): int64
-                  {.tags: [WriteDbEffect], raises: [].} =
```
Comment (dom96) [inline lib/impure/db_sqlite.nim:645]: Instead of removing this `raises`, maybe better to explicitly write out a list of exceptions that can be raised?

**Review by dom96 — APPROVED** [review]
Looks good to me.

As a side-note: these db modules are becoming dated and seem like great candidates for being moved out of the stdlib.

File: lib/impure/db_sqlite.nim (hunk 2 above)
Comment (ringabout) [reply lib/impure/db_sqlite.nim:645]: Yeah, I'm not sure about this. Maybe we should turn raise errors into warnings first to reduce breaking changes. 

I will add it anyway.

Comment (ringabout) [reply lib/impure/db_sqlite.nim:645]: solved

File: tests/stdlib/tsqlitebindatas.nim [hunk 3]
```diff
@@ -48,3 +48,25 @@ block tsqlitebindatas: ## db_sqlite binary data
...
+      removeFile("db.sqlite3")
+
+  block:
+    var db = db_sqlite.open("db.sqlite3", "", "", "")
+    try:
+      db.exec(sql("CREATE TABLE table1 (url TEXT, other_field INT);"))
+      db.exec(sql("INSERT INTO table1 (url, other_field) VALUES (?, ?);"), "http://domain.com/test?param=1", 123)
+    finally:
+      db.close()
+      removeFile("db.sqlite3")
```
Comment (timotheecour) [inline tests/stdlib/tsqlitebindatas.nim]: https://github.com/timotheecour/Nim/issues/534#issuecomment-897038592

File: tests/stdlib/tsqlitebindatas.nim [hunk 4]
```diff
@@ -48,3 +48,25 @@ block tsqlitebindatas: ## db_sqlite binary data
...
+
+
+block:
+  block:
+    var db = db_sqlite.open("db.sqlite3", "", "", "")
+    try:
+      db.exec(sql("CREATE TABLE table1 (url TEXT, other_field INT);"))
+      db.exec(sql("REPLACE INTO table (url, another_field) VALUES (?, '123');"))
+    except DbError as e:
+      doAssert e.msg == "The number of parameter doesn't match that of \"?\" "
```
Comment (timotheecour) [inline tests/stdlib/tsqlitebindatas.nim]: this test doesn't actually ensure that the exception is raised; you need a "witness" variable to ensure this, called after the whole block
```nim
var witness = false
...
   witness = true

doAssert witness
```

File: lib/std/private/dbutils.nim [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+  var res = ""
+  var a = 0
+  for c in items(string(formatstr)):
+    if c == '?':
+      if a == args.len:
+        dbError("""The number of parameter doesn't match that of "?" """)
```
Comment (timotheecour) [inline lib/std/private/dbutils.nim]: show the minimum sufficient runtime context in error msgs; eg:
```nim
dbError("The number of parameter doesn't match that of `?` " % [...])
```

Comment (Varriount) [reply lib/std/private/dbutils.nim]:
```suggestion
        dbError("""The number of query arguments given exceeds the number of parameters present in the query. """)
```
An even better error message would be to show how many arguments and how many parameters are present.

File: tests/stdlib/tsqlitebindatas.nim (hunk 3 above)
Comment (ringabout) [reply tests/stdlib/tsqlitebindatas.nim]: thanks

File: tests/stdlib/tsqlitebindatas.nim (hunk 4 above)
Comment (ringabout) [reply tests/stdlib/tsqlitebindatas.nim]: true

File: lib/std/private/dbutils.nim (hunk 5 above)
Comment (ringabout) [reply lib/std/private/dbutils.nim]: Positive.

> dbError("""The number of query arguments given exceeds the number of parameters present in the query. """)

It should be good enough for this super corner case.

File: tests/stdlib/tsqlitebindatas.nim (hunk 3 above)
Comment (ringabout) [reply tests/stdlib/tsqlitebindatas.nim]: thanks

**Review by timotheecour — APPROVED** [review]

**Comment by timotheecour** [comment]
added followup label for https://github.com/nim-lang/Nim/pull/18669/files#r686051480
